client/pool.c

```diff
@@ -971,13 +984,21 @@ static int send_data_to_connection(connection_list_element *connection, int *pro
 
 void *pool_main_thread(void *arg)
 {
+	pthread_cleanup_push(thread_finish_routine, &g_pool_main_thread_status);
+	pthread_mutex_lock(&g_pool_thread_status_mutex);
+	atomic_store_explicit_int(&g_pool_main_thread_status, THREAD_CANCELLABLE, memory_order_relaxed);
+	pthread_mutex_unlock(&g_pool_thread_status_mutex);
+	int old_state_type;
+	pthread_setcanceltype(PTHREAD_CANCEL_ASYNCHRONOUS, &old_state_type);
```

Comment:
Maybe use 
pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, &old_state_type);
pthread_setcanceltype(PTHREAD_CANCEL_DEFERRED, &old_state_type);
and use pthread_testcancel(pthread_self()) to set cancel point. 

And use
volatile thread_state  g_pool_main_thread_status instead of atomic operations since only one thread to modify, the other one to query, it can be simpler and more efficient.

Reply:
> Maybe use
> pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, &old_state_type);
> pthread_setcanceltype(PTHREAD_CANCEL_DEFERRED, &old_state_type);
> and use pthread_testcancel(pthread_self()) to set cancel point.

i should anyway add 
pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, &old_state_type); and pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, &old_state_type); to protect critical sections (to avoid that it will exit without releasing a mutex, for example) so basically the code would be the same as now but with PTHREAD_CANCEL_DEFERRED instead of PTHREAD_CANCEL_ASYNC, so i don't see any benefit to use PTHREAD_CANCEL_DEFERRED in this case, it would be just slightly slower to terminate.

about atomic, it is needed because of (for example) this case:
pool_main_thread write status THREAD_CANCELLABLE
if at the same time or some microseconds after it another thread want to terminate pool_main_thread, it may still read THREAD_INIT and shut down xdag without wait

and if i use the mutex in the the finish_routine there can be a deadlock (so i would need two mutexes)